### PR TITLE
EffSecShoot

### PR DESCRIPTION
**Pull request by TheAbsolutionism**
### Description
This PR aims to convert EffShoot into an EffSecShoot, allowing users to manipulate all or any entities/projectiles and shooters instantaneously

---
**Target Minecraft Versions:** any 
**Requirements:** none 
**Related Issues:** #4789

### CODE REVIEW

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+
+		return super.walk(event, false);
+	}
+
+	@SuppressWarnings("unchecked")
+	private static <E extends Entity> void set(Entity entity, EntityData<E> entityData) {
+		entityData.set((E) entity);
+	}
+
+	private Consumer<? extends Entity> getConsumer(Event event, Vector vector, EntityData<?> entityData, @Nullable LivingEntity shooter) {
```
[the remark this answers is not in the document]
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/sections/EffSecShoot.java:167]: mmm how about `getSpawnConsumer` or `getShootEventConsumer`?

File: src/main/java/ch/njol/skript/expressions/ExprShooter.java [hunk 2]
```diff
@@ -46,16 +47,24 @@ public class ExprShooter extends PropertyExpression<Projectile, LivingEntity> {
...
 	}
-	
+
+	private boolean withinShootEvent;
+
 	@SuppressWarnings({"unchecked", "null"})
 	@Override
 	public boolean init(final Expression<?>[] exprs, final int matchedPattern, final Kleenean isDelayed, final ParseResult parseResult) {
+		if (getParser().isCurrentEvent(EffSecShoot.ShootEvent.class) && exprs[0] == null)
+			withinShootEvent = true;
```
[the remark this answers is not in the document]
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/expressions/ExprShooter.java:57]: You think I should just only check for `instanceof ShootEvent` inside the get method?

File: src/test/skript/tests/syntaxes/sections/EffSecShoot.sk [hunk 3]
```diff
@@ -0,0 +1,33 @@
+test "EffSecShoot":
+	set {_loc} to spawn of world "world"
+	spawn pig at {_loc}:
+		set ai of entity to false
+		set {_shooter} to entity
+
+	shoot a pig from {_shooter}
+	assert last shot entity is a pig with "Last shot entity is not a pig"
```
Comment (sovdeeth) [inline src/test/skript/tests/syntaxes/sections/EffSecShoot.sk:8]: this pig is never deleted

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprShooter.java [hunk 4]
```diff
@@ -46,42 +29,46 @@ public class ExprShooter extends PropertyExpression<Projectile, LivingEntity> {
...
+			if (shooter instanceof LivingEntity livingShooter)
+				return livingShooter;
 			return null;
 		});
 	}
 	
 	@Override
 	@Nullable
-	public Class<?>[] acceptChange(final ChangeMode mode) {
+	public Class<?>[] acceptChange(ChangeMode mode) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprShooter.java:56]:
```suggestion
	public @Nullable Class<?>[] acceptChange(ChangeMode mode) {
```

File: src/main/java/ch/njol/skript/expressions/ExprShooter.java [hunk 5]
```diff
@@ -46,42 +29,46 @@ public class ExprShooter extends PropertyExpression<Projectile, LivingEntity> {
...
 		return super.acceptChange(mode);
 	}
 	
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) {
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
 		if (mode == ChangeMode.SET) {
 			assert delta != null;
-			for (final Projectile p : getExpr().getArray(e)) {
+			for (Projectile p : getExpr().getArray(event)) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprShooter.java:66]:
```suggestion
			for (Projectile projectile : getExpr().getArray(event)) {
```

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+		return super.walk(event, false);
+	}
+
+	@SuppressWarnings("unchecked")
+	private static <E extends Entity> void set(Entity entity, EntityData<E> entityData) {
+		entityData.set((E) entity);
+	}
+
+	private Consumer<? extends Entity> getSpawnConsumer(Event event, Vector vector, EntityData<?> entityData, @Nullable LivingEntity shooter) {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:176]: i feel the method name should actually include what the consumer does, not that it's a consumer cuz that's visible in the return type. something like `afterSpawn`

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+import org.bukkit.projectiles.ProjectileSource;
+import org.bukkit.util.Vector;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.util.List;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.function.Consumer;
+
+public class EffSecShoot extends EffectSection {
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:29]: missing docs

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 8]
```diff
@@ -0,0 +1,213 @@
...
+			if (delayed.get()) {
+				Skript.error("Delays cannot be used within a 'shoot' effect section");
+				return false;
+			}
+		}
+		return true;
+	}
+
+	@Override
+	@SuppressWarnings({"unchecked","rawtypes"})
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:115]: only suppress on the lines that actually need it

**Review by sovdeeth — CHANGES_REQUESTED** [review]
almost done

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 9]
```diff
@@ -0,0 +1,217 @@
...
+				if (shooter instanceof LivingEntity livingShooter) {
+					vector = finalDirection.getDirection(livingShooter.getLocation()).multiply(finalVelocity.doubleValue());
+					Class<? extends Entity> type = entityData.getType();
+					if (Fireball.class.isAssignableFrom(type)) {
+						if (trigger != null) {
+							//noinspection unchecked,rawtypes
+							livingShooter.getWorld().spawn(
+								livingShooter.getEyeLocation().add(vector.clone().normalize().multiply(0.5)),
+								type,
+								(Consumer) afterSpawn(event, vector, entityData, livingShooter)
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:135]: you could pull out all the afterSpawn()s and just make a local var at the instanceof LivingEntity check, meaning less duplicate code and fewer warnings to suppress

**Review by sovdeeth — APPROVED** [review]

**Comment by TheAbsolutionism** [comment]
Note:
For MC 1.19.4, to use launchProjectile with a consumer, requires the Bukkit Consumer type. Thus resulting the usage of Reflection in order to use the proper method.

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 10]
```diff
@@ -0,0 +1,264 @@
...
+			return shooter;
+		}
+
+		@Override
+		public @NotNull HandlerList getHandlers() {
+			throw new IllegalStateException();
+		}
+	}
+
+	private static final boolean RUNNING_1_20_4 = Skript.isRunningMinecraft(1, 20, 4);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:70]: what about 1.20.1?

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 11]
```diff
@@ -0,0 +1,264 @@
...
+                projectile.setShooter(shooter);
+                set(projectile, entityData);
+            }
+            ShootEvent shootEvent = new ShootEvent(entity, shooter);
+            lastSpawned = entity;
+            Variables.setLocalVariables(shootEvent, Variables.copyLocalVariables(event));
+            TriggerItem.walk(trigger, shootEvent);
+            Variables.setLocalVariables(event, Variables.copyLocalVariables(shootEvent));
+            Variables.removeLocals(shootEvent);
+        };
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:254]: indentation messed up

**Review by sovdeeth — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprShooter.java [hunk 12]
```diff
@@ -46,42 +29,45 @@ public class ExprShooter extends PropertyExpression<Projectile, LivingEntity> {
 	static {
 		Skript.registerExpression(ExprShooter.class, LivingEntity.class, ExpressionType.SIMPLE, "[the] shooter [of %projectile%]");
 	}
-	
+
 	@SuppressWarnings({"unchecked", "null"})
```
Comment (Efnilite) [inline src/main/java/ch/njol/skript/expressions/ExprShooter.java:33]: per statement

**Review by Efnilite — CHANGES_REQUESTED** [review]

File: src/main/java/ch/njol/skript/expressions/ExprShooter.java [hunk 13]
```diff
@@ -46,42 +29,45 @@ public class ExprShooter extends PropertyExpression<Projectile, LivingEntity> {
...
 	@Override
-	public void change(final Event e, final @Nullable Object[] delta, final ChangeMode mode) {
+	public void change(Event event, Object @Nullable [] delta, ChangeMode mode) {
 		if (mode == ChangeMode.SET) {
 			assert delta != null;
-			for (final Projectile p : getExpr().getArray(e)) {
-				assert p != null : getExpr();
-				p.setShooter((ProjectileSource) delta[0]);
+			for (Projectile projectile : getExpr().getArray(event)) {
+				assert projectile != null : getExpr();
```
[the remark this answers is not in the document]
Comment (TheAbsolutionism) [reply src/main/java/ch/njol/skript/expressions/ExprShooter.java:66]: Now that you mention it, I honestly dont know, I had just changed the variable name.

[the remark this answers is not in the document]
Comment (Efnilite) [reply src/main/java/ch/njol/skript/expressions/ExprShooter.java:66]: ah its like the message if the assert fails

**Review by Efnilite — APPROVED** [review]

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+	@Override
+	protected @Nullable TriggerItem walk(Event event) {
+		lastSpawned = null;
+		Number finalVelocity = velocity != null ? velocity.getSingle(event) : DEFAULT_SPEED;
+		Direction finalDirection = direction != null ? direction.getSingle(event) : Direction.IDENTITY;
+		if (finalVelocity == null || finalDirection == null)
+			return null;
+
+		for (Object shooter : shooters.getArray(event)) {
+			for (EntityData<?> entityData : types.getArray(event)) {
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:141]: need to get entitydatas outside of the loops, otherwise it'll be evaluated multiple times

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+					}
+				} else {
+					vector = finalDirection.getDirection((Location) shooter).multiply(finalVelocity.doubleValue());
+					if (trigger != null) {
+						//noinspection unchecked,rawtypes
+						entityData.spawn((Location) shooter, (Consumer) afterSpawn(event, entityData, null));
+					} else {
+						finalProjectile = entityData.spawn((Location) shooter);
+					}
+				}
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:220]: this is pretty hard to parse because of all the nested ifs
can you rework it using guard clause-like formatting, or with helper methods? It's just a bit hard to read with 7 levels of indentation

**Review by sovdeeth — CHANGES_REQUESTED** [review]
looking good

**Review by sovdeeth — APPROVED** [review]
EffSecShoot#walk() still isn't my favorite, but as discussed on discord it's good enough

File: src/main/java/ch/njol/skript/expressions/ExprShooter.java [hunk 16]
```diff
@@ -46,42 +29,45 @@ public class ExprShooter extends PropertyExpression<Projectile, LivingEntity> {
...
+				projectile.setShooter((ProjectileSource) delta[0]);
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/expressions/ExprShooter.java:67]: Rather than extracting and casting this for every element of the array it would be easier to just take it as a variable before the loop.

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 17]
```diff
@@ -0,0 +1,298 @@
...
+						case PROJECTILE_WORLD_NO_TRIGGER -> {
+							Projectile projectile = (Projectile) livingShooter.getWorld().spawn(shooterLoc, type);
+							projectile.setShooter(livingShooter);
+							finalProjectile = projectile;
+						}
+						case PROJECTILE_WORLD_TRIGGER -> {
+							//noinspection unchecked
+							livingShooter.getWorld().spawn(shooterLoc, type, afterSpawn);
+						}
+						default -> throw new IllegalStateException("Unexpected value: " + caseUsage);
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:209]: This case usage stuff seems kind of messy and this method is becoming a bit too long and complex. The cyclomatic complexity is ridiculously high (somewhere over 50 by my headcount) and I think it needs to be broken up.

You've got a few options (e.g. move part of the method from the loop into its own method) but personally I would recommend that you make the CaseUsage a functional interface (accepting the required variables) and then you can move each application into its own space and eliminate the switch entirely.

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 18]
```diff
@@ -0,0 +1,298 @@
...
+			Variables.setLocalVariables(event, Variables.copyLocalVariables(shootEvent));
+			Variables.removeLocals(shootEvent);
+		};
+	}
+
+	@Override
+	public String toString(@Nullable Event event, boolean debug) {
+		return "shoot " + types.toString(event, debug) + " from " + shooters.toString(event, debug) +
+			(velocity != null ? " at speed " + velocity.toString(event, debug) : "") +
+			(direction != null ? " " + direction.toString(event, debug) : "");
```
Comment (Moderocky) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:295]: When you're appending strings over multiple lines, the pluses should go at the start of the new line.
```suggestion
		return "shoot " + types.toString(event, debug) + " from " shooters.toString(event, debug) +
			+ (velocity != null ? " at speed " + velocity.toString(event, debug) : "")
			+ (direction != null ? " " + direction.toString(event, debug) : "");
```
This convention comes from mathematics because it makes it easier to determine the relation between elements in a displayed formula. If you're interested to learn about it, Donald Knuth explains it in _The TeXBook_ (pp. 195-196).

**Review by Moderocky — CHANGES_REQUESTED** [review]
The idea is nice but the implementation needs some work.

**Review by Moderocky — APPROVED** [review]

File: src/main/java/ch/njol/skript/sections/EffSecShoot.java [hunk 19]
```diff
@@ -46,13 +46,61 @@ public class EffSecShoot extends EffectSection {
...
+		PROJECTILE_WORLD_TRIGGER {
+			@Override
+			public @Nullable Entity shootHandler(EntityData<?> entityData, LivingEntity shooter, Location location, Class<? extends Entity> type, Vector vector, Consumer<?> consumer) {
+				//noinspection unchecked,rawtypes
+				shooter.getWorld().spawn(location, type, (Consumer) consumer);
+				return null;
+			}
+		};
+
+		public abstract @Nullable Entity shootHandler(EntityData<?> entityData, LivingEntity shooter, Location location, Class<? extends Entity> type, Vector vector, Consumer<?> consumer);
```
Comment (sovdeeth) [inline src/main/java/ch/njol/skript/sections/EffSecShoot.java:103]: should have docs

**Review by sovdeeth — APPROVED** [review]
